### ISSUE TITLE
Add support to execute command after login through ssh

### ISSUE THREAD

**Issue by bharat555** [mRemoteNG/mRemoteNG#503]
If possible support for sending custom keystrokes(commands) after login through ssh would be great. This could be useful to add temporary shell aliases/functions or for example to change to one's preferred shell after login.

More like "External Tools Before" but the keystrokes should be sent to the terminal instead of an external program.

**Comment by bharat555** [comment]
@thomasjammet this is helpful, but that requires creating a session profile for each connection in Putty. Managing that option from within mRemoteNG connection config will be better.

**Comment by lwuthunder** [comment]
This is the most wanted feature for mRemoteNG. Can we have two new fields in config after user name and password, one for login path, one for login command? That will allow quick config and very useful.

**Comment by esturniolo** [comment]
I vote for this too.

**Comment by sergiimatus** [comment]
it's must have feature, as many rivals already has it

**Comment by maxexcloo** [comment]
+1 vote for this!

**Comment by gavu** [comment]
+1

**Comment by supertino7** [comment]
+1

**Comment by acchr** [comment]
+1

**Comment by ktp1976** [comment]
I don't think that one command for all PuTTY default sessions is sufficient. I am used PAC Manager on Linux (https://sourceforge.net/projects/pacmanager/) and there you can configure Tcl EXPECT separately for every connection to read the console output and perform actions based on that output. I am badly missing this feature with Windows and mRemote! If possible have a look and try to implement such features. This is a must have for example to connect to a jump host and from there connect to other systems.

**Comment by intermike** [comment]
+1

**Comment by oallauddin** [comment]
+1 We have multiple environments on one server. In mRemoteNG I have a connection for each environment. Options discussed by @lwuthunder would be ideal. On connection the paths and environment variables needed for the environment would be already set up.

**Comment by ron4all** [comment]
+1
Would be great, if it could be
* an (inheritable) text-input-field to enter,
* any number of lines.

The "5 Commands"-Limit is somehow unhandy when using Putty-CM.

**Comment by ananttickoo** [comment]
+1
this would be helpful in case we have multiple jumpbox to reach destination. If we can send text to window it would be helpful. 
e.g custom commands, complex passwords.

**Comment by jornlux** [comment]
+1 vote for this!

**Comment by ericmck2000** [comment]
+1. MTPutty has it. This needs it.

**Comment by matejsp** [comment]
@ericmck2000 Tnx for the tip ... really appreciated!!! Didn't know that such tool existed.

**Comment by bharat555** [comment]
> 
> 
> +1. MTPutty has it. This needs it.

This is exactly what I was looking for, After approx 2 years of waiting.

**Comment by frelis** [comment]
+1 for this functionality.

**Comment by R0GGER** [comment]
+1

**Comment by jacquen** [comment]
+ 1

**Comment by mosyae** [comment]
+1

**Comment by k00lith** [comment]
+1

**Comment by karthikperu7** [comment]
+1. Came from MTputty. This stopped me from continuing with mRemoteNG.

**Comment by mstrullu** [comment]
+1

**Comment by grego1981** [comment]
+1

**Comment by EdoardoVarani** [comment]
+1

**Comment by EXTRMKAISER** [comment]
+1

**Comment by bender270** [comment]
+1

**Comment by aazizitarik** [comment]
+1

**Comment by Trans-iX** [comment]
+1 Would really like to see a Login Macro of Password Macro. We always login with a unprivileged and then su - to root with a password. Remote Desktop Manager has this "Macro password" option.

**Comment by csalles1** [comment]
+1

**Comment by ktp1976** [comment]
I'm giving up on this. My company is going to switch to MobaXterm anyways. "MobaXterm supports code snippets that get recorded as a macro and can get replayed on other servers later. Besides that you have multiexecution of typed commands, that is the commands get executed on different servers at the same time."
For private use I stick with PAC Manager on Linux which uses expect for macros.
So Long, and Thanks for All the Fish.

**Comment by LorisAlbanese** [comment]
+1 for this feature but ... menwhile ... i switched to "multi putty manager",  lighter then Moba.

**Comment by fcolak** [comment]
+1 This is definitely needed..

**Comment by xjikka** [comment]
+1 Yeah, for me too, pretty pretty please with sugar on the top ;-)

**Comment by R0GGER** [comment]
I switched to [Royal TS](https://www.royalapps.com/ts/win/features) (it's cheap) and has all features mRemoteNG does not have and won't ever get.

**Comment by vidyasagarreddyc** [comment]
+1

**Comment by merarischroeder** [comment]
This issue is duplicated by https://github.com/mRemoteNG/mRemoteNG/issues/1855

**Comment by merarischroeder** [comment]
I am looking at this myself to check the feasibility. I haven't worked on this project before. I have cloned the C# code and had a look around.

It looks like a Putty window is used, and mRemoteNG will need to send the keystrokes to the window handle at the right time (after a delay).

Later when SSH.NET is used, it will be easier to integrate.

**Comment by merarischroeder** [comment]
Work in progress - https://github.com/mRemoteNG/mRemoteNG/pull/2064

**Comment by Boscop** [comment]
@thomasjammet @genettistylez Thanks but your Putty workaround doesn't work for me. I followed your steps, but now, whenever I connect by double clicking on a SSH connection, after I enter my password, the tab closes so the connection was aborted. Any idea why? :)
Btw, the command I had entered in Putty config is `tmux new -A -s ssh_tmux`. (And I had previously confirmed that it works when I enter it manually after connecting.)
I'd appreciate any info on how I can make it work, I'm using the latest stable version btw.

**Comment by Kvarkas** [comment]
not relevant

**Comment by R0GGER** [comment]
An issue that’s been open since 2017 is suddenly no longer relevant, without any explanation why…?? Unbelievable.

**Comment by Kvarkas** [comment]
Have you tryed NB version? Its already implemented 🤷‍♂️

**Comment by heitmanr** [comment]
latest release hasn't been changed. Could you explain what's the NB Version?

nightly build, my bad ;-)

so MremoteNG is Next Generation again? ;-)

**Comment by bharat555** [comment]
I requested this feature in 2017, and completely lost interest in this topic, thanks for reviving this thread and much thanks to @merarischroeder for implementing it ( https://github.com/mRemoteNG/mRemoteNG/issues/503#issuecomment-967223864 ) . I am yet to test it, but good to know that it has been implemented finally.

**Comment by Kvarkas** [comment]
> latest release hasn't been changed. Could you explain what's the NB Version?
> nightly build, my bad ;-)

Yes, due swith to latest .NET 9 (before release perhaps will switch to .net 10 - just for longer support etc)

> so MremoteNG is Next Generation again? ;-)

:) kind of :)  I prefer to shorter it to just "mR" but its always be mRemoteNG :) 

due natural improvements and progress support older versions is mission imposible, so instead i focused to make NB -> RC (Release Candidate) and after testing release as stable version so all could update and this will be supported long term as well.

**Comment by Kvarkas** [comment]
> I requested this feature in 2017, and completely lost interest in this topic, thanks for reviving this thread and much thanks to its was

It has been dark ages for mRemoteNG as old team lost interest in developing and prefer take theys time with family or work or something else, we dont blame them for what! But as for now I am only one with few entusiasts (and few AIs) who keep developing and improving mRemoteNG to return back to its Glory :)

**Comment by pabcgitadmin** [comment]
Good to see at least few persons still interested in this great project.
